ostd/src/mm/frame/allocator.rs

```diff
@@ -140,73 +125,207 @@ fn test_alloc_dealloc() {
+                }
             }
-            None => None,
-        }
-    }
 
-    // TODO: this method should be marked unsafe as invalid arguments will mess
-    // up the underlying allocator.
-    pub fn dealloc(&mut self, start_frame: usize, count: usize) {
-        self.allocator.dealloc(start_frame, count);
-        self.allocated -= count * PAGE_SIZE;
+            fn add_free_frames(range: Range<Paddr>) {
+                log::info!("Adding free frames to the allocator: {:x?}", range);
+                dealloc_upcall(range.start, range.len());
+            }
+        }
     }
+}
 
-    pub fn mem_total(&self) -> usize {
-        self.total
-    }
+/// An allocator in the early boot phase when frame metadata is not available.
+pub(super) struct EarlyFrameAllocator {
+    under_4g_region: MemoryRegion,
+    under_4g_end: Paddr,
 
-    pub fn mem_available(&self) -> usize {
-        self.total - self.allocated
-    }
+    max_region: Option<MemoryRegion>,
+    max_end: Option<Paddr>,
 }
 
-pub(in crate::mm) static FRAME_ALLOCATOR: Once<SpinLock<CountingFrameAllocator>> = Once::new();
+pub(super) static EARLY_ALLOCATOR: spin::Mutex<Option<EarlyFrameAllocator>> =
+    spin::Mutex::new(None);
 
-pub(crate) fn init() {
-    let regions = &crate::boot::EARLY_INFO.get().unwrap().memory_regions;
-    let mut total: usize = 0;
-    let mut allocator = FrameAllocator::<32>::new();
-    for region in regions.iter() {
-        if region.typ() == MemoryRegionType::Usable {
-            // Make the memory region page-aligned, and skip if it is too small.
-            let start = region.base().align_up(PAGE_SIZE) / PAGE_SIZE;
-            let region_end = region.base().checked_add(region.len()).unwrap();
-            let end = region_end.align_down(PAGE_SIZE) / PAGE_SIZE;
-            if end <= start {
+impl EarlyFrameAllocator {
+    /// Creates a new early frame allocator.
+    ///
+    /// It uses at most 2 regions, the first is the maximum usable region below
+    /// 4 GiB. The other is the maximum usable region above 4 GiB and is only
+    /// usable when linear mapping is constructed.
+    pub fn new() -> Self {
+        let regions = &crate::boot::EARLY_INFO.get().unwrap().memory_regions;
+        let mut max_u4g_size = 0;
+        let mut max_u4g_index = 0;
+
+        let mut max_size = 0;
+        let mut max_index = 0;
+        for (i, region) in regions.iter().enumerate() {
+            if region.typ() != MemoryRegionType::Usable {
                 continue;
             }
-            // Add global free pages to the frame allocator.
-            allocator.add_frame(start, end);
-            total += (end - start) * PAGE_SIZE;
-            info!(
-                "Found usable region, start:{:x}, end:{:x}",
-                region.base(),
-                region.base() + region.len()
+            const PADDR4G: Paddr = 0x1_0000_0000;
+            if region.base() < PADDR4G && region.len() > max_u4g_size {
```

Comment:
`end() <= PADDR4G` in case there is a memory region that crosses the 4G boundary.

In practice, this won't happen, but since we never check this property, we'd better have more robust code to handle more situations.